File: Codeforces/compfest/M1.cpp

#include <vector>
#include <set>
#include <iostream>
#include <algorithm>
#include <deque>

using namespace::std;


typedef unsigned long long ull;

struct edge
{
	ull a;
	ull b;
	ull cost;

	edge(ull a, ull b, ull cost) : a(a), b(b), cost(cost) {}
	edge() : a(), b(), cost() {}
};

const ull INF = (int)1e18;

struct node
{
	ull n;
	ull forward = INF;
	ull backward = INF;
	bool visited = false;

	node(ull n) : n(n) {}
};

void bfs_forward(deque<pair<node *, ull>> &q, const vector<vector<pair<node *, ull>>> &inc)
{
	while (true)
	{
		pair<node *, ull> p = q.front();
		q.pop_front();
		node *n = p.first;
		ull c = p.second;
		n->visited = true;
		for (pair<node *, ull> pp : inc.at(n->n))
		{
			node *e = pp.first;
			e->forward = min(e->forward, n->forward + c);
			if (e->visited)
				continue ;
			q.push_back(pp);
		}
	}
}


int main()
{
	int n, m;
	cin >> n >> m;

	vector<edge> v(m);
	for (int i=0; i < m; ++i)
	{
		int a, b, c;
		cin >> a >> b >> c;
		--a;
		--b;
		v.at(i) = edge(a, b, c);
	}
	
	sort(v.begin(), v.end(), [](const edge &a, const edge &b){ return a.cost < b.cost; });
	vector<vector<pair<node *, ull>>> inc(n);
	vector<node *> nodes(n);
	for (int i=0; i < n; ++i)
		nodes.at(i) = new node(i);
	for (const auto &i : v)
		inc.at(i.a).push_back({nodes.at(i.b), i.cost});
	
	deque<pair<node *, ull>> q;
	q.push_back(pair<node *, ull>(nodes.at(0), inc.at(0).at(0).second));
	bfs_forward(q, inc);

	for (const node *n : nodes)
		cout << n->forward << ' ';
	cout << endl;
}

File: Codeforces/compfest/M1.py

from collections import deque

INF = int(1e8)


class node():
    def __init__(self, n):
        self.n = n
        self.forward = INF
        self.backward = INF
        self.visited = 0

    def __repr__(self):
        return 'node{' f'{self.n+1, self.forward, self.backward, self.visited}' '}'

    def __str__(self):
        return self.__str__()


class edge():
    def __init__(self, a, b, cost):
        self.a = a
        self.b = b
        self.cost = cost

    def __repr__(self):
        return 'EDGE{' f'{self.a, self.b} cost<{self.cost}>'

    def __str__(self):
        return self.__repr__()


def inp():
    return list(map(int, input().split()))


def bfs_fwd():
    while True:
        if len(q) == 0:
            break
        front = q.popleft()
        front.visited = 1
        for nd in inc[front.n]:
            ed = edges.get((front.n, nd.n), INF)
            nd.forward = min(nd.forward, front.forward + ed)
            if nd.visited:
                continue
            q.append(nodes[nd.n])


def bfs_bwd():
    while True:
        if len(q) == 0:
            break
        front = q.popleft()
        front.visited = 1
        for nd in inc_back[front.n]:
            ed = edges.get((nd.n, front.n), INF)
            nd.backward = min(nd.backward, front.forward + ed, front.backward + ed, nd.forward)
            if nd.visited:
                continue
            q.append(nodes[nd.n])


n, m = inp()

nodes = [0] * n
for i in range(n):
    nodes[i] = node(i)

v = [0] * m
for i in range(m):
    a, b, c = inp()
    a -= 1
    b -= 1
    v[i] = edge(a, b, c)

v.sort(key=lambda x: x.cost)

edges = {(e.a, e.b): e.cost for e in v}
nodes[0].forward = 0
nodes[0].backward = 0

inc = [[] for _ in range(n)]
for i in v:
    inc[i.a].append(nodes[i.b])

inc_back = [[] for _ in range(n)]
for i in v:
    inc_back[i.b].append(nodes[i.a])

q = deque()
q.append(nodes[0])
bfs_fwd()

for _n in nodes:
    _n.visited = 0

q.append(nodes[0])
bfs_bwd()

# print(*inc, sep='\n')
# print()
# print()
# print(nodes)
#
for i in range(1, n):
    b = nodes[i].backward
    if b == INF:
        print(-1, end=' ')
    else:
        print(b, end=' ')
print()
